### ISSUE TITLE
`model_code` argument

### ISSUE THREAD

**Issue by vincentarelbundock** [stan-dev/cmdstanr#252]
Hi,

Thanks for your great work on `cmdstanr`. I really appreciate the awesome tools you make available!

I feel it would be convenient to write Stan code directly as a string inside my `.R` scripts. The `rstan` package offers a `model_code` argument which allows users to pass models as strings. Here is the doc for that argument:

> model_code
> A character string either containing the model definition or the name of a character string object in the workspace. This argument is used only if arguments file and fit are not specified.

Of course, I could write a function to `writeLines` a `model.stan` file to disk, and then run `cmdstan_model` on the file, but my attempts at this feel very hack-ish. An "official" solution would be great!

Do you think there would be appetite for this in the community?

**Comment by mitzimorris** [comment]
My answer to this request is always "NO", for the following reasons:

- CmdStan doesn't accept strings as inputs; therefore the CmdStanX wrapper interfaces don't either. 

-  If your model is so trivial that you can type it in correctly as a string, you would be better served using either RStanArm or BRMS which will probably do what you want, and perhaps use better priors, standardize inputs, etc.

- If you are developing a model for a serious analysis of some dataset, then you will need to include the Stan code as part of the writeup, paper, reproducable workflow, etc.

- If you're working in RStudio, there's a Stan editor available, so there's really no barrier to writing a Stan program and saving it to a file.

**Comment by rok-cesnovar** [comment]
@vincentarelbundock Thanks for the feature request. 

A model_code arg to cmdstan_model is not in our plans, but we have an "official" solution for that which is via write_stan_tempfile. See https://mc-stan.org/cmdstanr/reference/write_stan_tempfile.html

**Comment by vincentarelbundock** [comment]
Sounds like this has been discussed many times already. I'm sorry I missed it; I'm not yet familiar with all the venues where such things are talked about.

Thanks for the thoughtful response @mitzimorris . My use-case is writing Stan tutorials for in Rmarkdown notebooks. So you're right, the models are "trivial" and not "serious". But for beginners, it's useful not to jump to RStanArm immediately, and it's also useful to display the model code in the same notebook as the rest of the code.

That said, the solution proposed by @rok-cesnovar seems perfectly satisfactory, as it should only add a single line of code to my notebooks. Excellent!

Thanks for your time and help!

**Comment by mitzimorris** [comment]
@rok-cesnovar  - very nice!  this could be extended to "string-to-stanfile" or something like that which would take 2 args - the string and the filename.  and if you clobber existing files, c'est la vie.  for this use case, it would work.

@vincentarelbundock - you use case is totally valid - not only that, Andrew Gelman made a very similar request - not to allow strings as inputs, but to have a way to do the kind of successive edits for teaching.  my attempts at a Jupyter notebooks are here:  https://github.com/stan-dev/example-models/tree/master/jupyter/covid-inf-rate.  I did a case study on Jupyter notebooks for Stan - https://mc-stan.org/users/documentation/case-studies/jupyter_colab_notebooks_2020.html,  all source code here: https://github.com/stan-dev/example-models/tree/master/jupyter/covid-inf-rate

you're welcome to contribute!  and if you or any of your students or collaborators like to hack Jupyter notebooks, it would be cool to develop a multi-pane notebook display which would allow for interactive edits in one window, and fitting the model in another.

also, apologies for my rather harsh response and thanks for your kind understanding - you're quite right that this has been requested many times, hence my lack of patience.  I spent a long time as a support programmer in biology labs.  the really good researchers insisted on strict attention to detail - follow the protocol and document every step, before and after details - in the lab notebook - and in those days (mid 2000s), the notebooks were literally notebooks.  in the software world, the use of revision control systems (git and all predecessors) is similarly necessary, even for single-author works.  hence my aversion to just shipping around data strings - it's teaching bad habits.
